src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php

```diff
@@ -298,19 +326,27 @@ private function extractFromReflectionType(\ReflectionType $reflectionType, \Ref
             $type = new Type(Type::BUILTIN_TYPE_NULL, $nullable);
         } elseif ($reflectionType->isBuiltin()) {
             $type = new Type($phpTypeOrClass, $nullable);
+        } elseif (null !== $reflectionMethod) {
+            $type = new Type(Type::BUILTIN_TYPE_OBJECT, $nullable, $this->resolveTypeName($phpTypeOrClass, $reflectionMethod, null));
+        } elseif (null !== $reflectionProperty) {
+            $type = new Type(Type::BUILTIN_TYPE_OBJECT, $nullable, $this->resolveTypeName($phpTypeOrClass, null, $reflectionProperty));
         } else {
-            $type = new Type(Type::BUILTIN_TYPE_OBJECT, $nullable, $this->resolveTypeName($phpTypeOrClass, $reflectionMethod));
+            throw new \BadMethodCallException('You should provide method or property reflection');
         }
 
         return $type;
     }
 
-    private function resolveTypeName(string $name, \ReflectionMethod $reflectionMethod): string
+    private function resolveTypeName(string $name, ?\ReflectionMethod $reflectionMethod, ?\ReflectionProperty $reflectionProperty): string
     {
+        if (null === ($reflection = $reflectionMethod ?? $reflectionProperty)) {
```

Comment:
calling the variable `$reflector`  would be better, as that's the naming used by PHP for its interface implemented by all reflection classes (but it does not make sense to use it as typehint, as `getDeclaringClass` is not part of this interface as other reflection classes don't implement it)